Extract message from the test error and print this instead of truncating the whole error object

### Clear and concise description of the problem

Hello :wave:

Many thanks for your work on vitest so far. I would like to request that errors printed by vitest are handled in a similar way to Jest, where the message is printed in red and the rest of the object omitted.

When certain debug flags are provided to the tests I would like to throw a descriptive error message in case they fail validation; and then depending on the result of the debug flag I can change how my internal logic behaves. This is to enable things like using a mock server vs calling the network whilst using the same test definitions.

In order to get this behaviour I am applying the following workaround:

```typescript
import { describe, expect, it } from 'vitest';

async function debug() {
    const trueValues = ['y', 'yes', 't', 'true', '1'];
    const falseValues = ['n', 'no', 'f', 'false', '0'];

    const allValues = [...trueValues, ...falseValues];

    const { DEBUG } = process.env;

    if (!DEBUG) {
        return true;
    }

    if (!allValues.includes(DEBUG)) {
        throw new Error(
            `Expected process.env.DEBUG to be one of ${allValues.join(
                ', ',
            )} but got "${DEBUG}" instead`,
        );
    }

    return trueValues.includes(DEBUG);
}

describe('debug', () => {
    it('does not truncate the error message using extra boilerplate', async () => {
        try {
            const result = await debug();
            expect(result).toEqual(true);
        } catch (error) {
            expect.fail(error.message);
        }
    });

    it('truncates the error message otherwise', async () => {
        await expect(debug()).resolves.toEqual(true);
    });
});
```

Run this with `DEBUG=never vitest run` in order to print the following:

```
 ❯ src/truncate.test.ts (2)
   ❯ debug (2)
     × does not truncate the error message using extra boilerplate
     × truncates the error message otherwise

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Failed Tests 2 ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

 FAIL  src/truncate.test.ts > debug > does not truncate the error message using extra boilerplate
AssertionError: Expected process.env.DEBUG to be one of y, yes, t, true, 1, n, no, f, false, 0 but got "never" instead
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/2]⎯

 FAIL  src/truncate.test.ts > debug > truncates the error message otherwise
AssertionError: promise rejected "Error: Expected process.env.DEBUG to be o…" instead of resolving
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[2/2]⎯

 Test Files  1 failed (1)
      Tests  2 failed (2)
```

### Suggested solution

Contrast this to Jest; when we run the second test we get the following output:

```
  debug
    ✕ truncates the error message otherwise (3 ms)

  ● debug › truncates the error message otherwise

    expect(received).resolves.toEqual()

    Received promise rejected instead of resolved
    Rejected to value: [Error: Expected process.env.DEBUG to be one of y, yes, t, true, 1, n, no, f, false, 0 but got "never" instead]

      35 |
      36 |     it('truncates the error message otherwise', async () => {
    > 37 |         await expect(debug()).resolves.toEqual(true);
         |               ^
      38 |     });
      39 | });
      40 |

      at expect (node_modules/expect/build/index.js:113:15)
      at Object.<anonymous> (src/truncate.test.ts:37:15)

Test Suites: 1 failed, 1 total
Tests:       1 failed, 1 total
```

### Alternative

I am using this workaround:

```typescript
try {
    const result = await debug();
    expect(result).toEqual(true);
} catch (error) {
    expect.fail(error.message);
}
```

This works but it isn't as nice. 😅 

### Additional context

I came across another workaround:

```
import {defineConfig} from "vitest/config";

export default defineConfig({
    test: {
        chaiConfig: {
            truncateThreshold: 120
        }
    }
});
```

When using this I get the following:

```
 FAIL  src/truncate.test.ts > debug > truncates the error message otherwise
AssertionError: promise rejected "Error: Expected process.env.DEBUG to be one of y, yes, t, true, 1, n, no, f, false, 0 but got "never" instead" instead of resolving
```

The problem I have with this is that it sets the threshold for all serialized objects, so if I want to increase the limit (for other purposes) I will get a lot of additional output for error objects too.

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

--- a/packages/expect/src/jest-expect.ts
+++ b/packages/expect/src/jest-expect.ts
@@ -675,7 +675,8 @@ export const JestChaiExpect: ChaiPlugin = (chai, utils) => {
               const _error = new AssertionError(
                 `promise rejected "${utils.inspect(err)}" instead of resolving`,
                 { showDiff: false },
-              )
+              ) as Error
+              _error.cause = err
               _error.stack = (error.stack as string).replace(error.message, _error.message)
               throw _error
             },
@@ -712,8 +713,8 @@ export const JestChaiExpect: ChaiPlugin = (chai, utils) => {
             (value: any) => {
               const _error = new AssertionError(
                 `promise resolved "${utils.inspect(value)}" instead of rejecting`,
-                { showDiff: false },
-              )
+                { showDiff: true, expected: new Error('rejected promise'), actual: value },
+              ) as any
               _error.stack = (error.stack as string).replace(error.message, _error.message)
               throw _error
             },

--- a/test/core/test/jest-expect.test.ts
+++ b/test/core/test/jest-expect.test.ts
@@ -814,6 +814,8 @@ describe('async expect', () => {
     }
     catch (err: any) {
       expect(err.message).toMatchInlineSnapshot(`"promise rejected "Error: some error { foo: { bar: 42 } }" instead of resolving"`)
+      expect(err.cause).toBeDefined()
+      expect(err.cause.message).toMatchInlineSnapshot(`"some error"`)
       expect(err.stack).toContain('jest-expect.test.ts')
     }
   })

